### PR TITLE
Convert Stable Diffusion ControlNet to TensorRT

### PR DESCRIPTION
**Pull request by dotieuthien**
Convert sd +  multicontrolnet to onnx 
Convert Unet to TensorRT

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @dotieuthien could you run `make style` to fix the tests?

**Comment by dotieuthien** [comment]
Thank you @patrickvonplaten, I did it

**Comment by dotieuthien** [comment]
Hi @patrickvonplaten, I have already add converting tensorrt for SD XL.
- Convert SD XL to onnx: `python convert_stable_diffusion_controlnet_to_onnx.py --model_path stabilityai/stable-diffusion-xl-base-1.0 --controlnet_path SargeZT/sdxl-controlnet-seg --output_path path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0 --fp16 --sd_xl`
- Convert SD XL onnx to tensorrt: `python convert_stable_diffusion_controlnet_to_tensorrt.py --onnx_path path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0/unet/model.onnx --output_path path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0/unet/model.engine --fp16 --num_controlnet 1 --sd_xl`

**Comment by dotieuthien** [comment]
Also, I create a simple test code to fast run unet tensorrt model to check the final result image here: https://github.com/dotieuthien/diffusers/blob/add-tensorrt-test/tests/pipelines/controlnet/test_controlnet_sdxl_v2.py

**Comment by dotieuthien** [comment]
image

**Review by sayakpaul — APPROVED** [review]
Looks good to me. 

I would highly suggest to add some comments at the beginning of both the scripts focusing on the following:

* What is the script for? 
* Example command to run the script successfully and expected outputs
* Dependencies needed to run the script 
* How to perform inference with the generated artifacts from the script (fine to just provide links) 

Let me know if anything is unclear.

**Comment by dotieuthien** [comment]
I have already updated comments, thank you @sayakpaul

File: scripts/convert_stable_diffusion_controlnet_to_onnx.py [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+    --controlnet_path SargeZT/sdxl-controlnet-seg
+    --output_path  path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0
+    --fp16
+    --sd_xl
+
+    Returns:
+        text_encoder/model.onnx
+        unet/model.onnx + unet/weights.pb
+        vae_encoder/model.onnx
+        vae_decoder/model.onnx
```
Comment (sayakpaul) [inline scripts/convert_stable_diffusion_controlnet_to_onnx.py:255]: How about also mentioning how to use these?

File: scripts/convert_stable_diffusion_controlnet_to_tensorrt.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    Example for SD XL:
+    python convert_stable_diffusion_controlnet_to_tensorrt.py
+    --onnx_path path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0/unet/model.onnx
+    --output_path path-to-models-stable_diffusion/stable-diffusion-xl-base-1.0/unet/model.engine
+    --fp16
+    --num_controlnet 1
+    --sd_xl
+
+    Returns:
```
Comment (sayakpaul) [inline scripts/convert_stable_diffusion_controlnet_to_tensorrt.py:26]: Would be great to also include something that shows how to use it.

File: scripts/convert_stable_diffusion_controlnet_to_onnx.py (hunk 1 above)
Comment (dotieuthien) [reply scripts/convert_stable_diffusion_controlnet_to_onnx.py:255]: Can I add script to test the onnx and trt model in this PR?

**Comment by sayakpaul** [comment]
We're working on a fix. Lets wait for a bit. Ping me after a few hours :)

**Comment by dotieuthien** [comment]
@nistvan86 Sorry I have not experienced ComfyUI yet, I will try later. You can refer my draft code to test SD XL with controlnet in the [comment](https://github.com/huggingface/diffusers/pull/4465#issuecomment-1668498081).

**Comment by dotieuthien** [comment]
@sayakpaul Could you please review my code?

**Comment by sayakpaul** [comment]
@dotieuthien let's not introduce any new things other than the conversion script and examples in this PR. The pipelines you are introducing are a separate PR candidate and should lie in `optimum` (cc: @echarlaix).

**Comment by dotieuthien** [comment]
Okay @sayakpaul , I have already moved all pipeline script to 2 examples, just keep the conversion script.

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
Thanks for your hard work!

**Comment by DIKODEY** [comment]
Hello, I'm not sure if I'm in the right place. I used this method https://community.amd.com/t5/ai/how-to-automatic1111-stable-diffusion-webui-with-directml/ba-p/649027 which works perfectly well, but I'm unable to use ControlNet. I'm not a professional, but I'm desperately looking for a solution. Do you have any ideas on how to proceed?

**Pull request merged**